File: blockchain_integration/pi_network/qrzk-mpc/node/node.c
// node.c

#include "node.h"
#include <openssl/aes.h>
#include <openssl/rsa.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

// Function to create a new node
Node* node_new(uint64_t id, RSA* public_key, RSA* private_key, uint64_t network_id) {
    Node* node = malloc(sizeof(Node));
    node->id = id;
    node->public_key = public_key;
    node->private_key = private_key;
    node->network_id = network_id;
    node->peers = hashmap_new();
    node->blockchain = malloc(sizeof(Block*));
    node->blockchain_size = 0;
    node->mempool = malloc(sizeof(Transaction*));
    node->mempool_size = 0;
    node->node_state = NODE_STATE_IDLE;
    return node;
}

// Function to add a peer to the node
void node_add_peer(Node* node, uint64_t peer_id, int socket) {
    hashmap_put(node->peers, peer_id, socket);
}

// Function to remove a peer from the node
void node_remove_peer(Node* node, uint64_t peer_id) {
    hashmap_remove(node->peers, peer_id);
}

// Function to send a message to a peer
void node_send_message(Node* node, uint64_t peer_id, char* message) {
    int socket = hashmap_get(node->peers, peer_id);
    if (socket!= -1) {
        AES_KEY aes_key;
        AES_set_encrypt_key(node->public_key, 256, &aes_key);
        uint8_t encrypted_message[1024];
        int encrypted_length = AES_encrypt(message, encrypted_message, &aes_key);
        send(socket, encrypted_message, encrypted_length, 0);
    }
}

// Function to receive a message from a peer
void node_receive_message(Node* node, char* message) {
    AES_KEY aes_key;
    AES_set_decrypt_key(node->private_key, 256, &aes_key);
    uint8_t decrypted_message[1024];
    int decrypted_length = AES_decrypt(message, decrypted_message, &aes_key);
    if (decrypted_length > 0) {
        if (strncmp(decrypted_message, "block", 5) == 0) {
            // Handle block message
            Block* block = malloc(sizeof(Block));
            memcpy(block, decrypted_message + 5, decrypted_length - 5);
            node->blockchain = realloc(node->blockchain, (node->blockchain_size + 1) * sizeof(Block));
            node->blockchain[node->blockchain_size] = block;
            node->blockchain_size++;
        } else if (strncmp(decrypted_message, "transaction", 12) == 0) {
            // Handle transaction message
            Transaction* transaction = malloc(sizeof(Transaction));
            memcpy(transaction, decrypted_message + 12, decrypted_length - 12);
            node->mempool = realloc(node->mempool, (node->mempool_size + 1) * sizeof(Transaction));
            node->mempool[node->mempool_size] = transaction;
            node->mempool_size++;
        } else {
            // Handle unknown message
            printf("Unknown message: %s\n", decrypted_message);
        }
    }
}

// Function to start the node
void node_start(Node* node) {
    int socket = socket(AF_INET, SOCK_STREAM, 0);
    struct sockaddr_in server_address;
    server_address.sin_family = AF_INET;
    server_address.sin_port = htons(8080);
    inet_pton(AF_INET, "0.0.0.0", &server_address.sin_addr);
    bind(socket, (struct sockaddr*)&server_address, sizeof(server_address));
    listen(socket, 3);

    while (1) {
        struct sockaddr_in client_address;
        socklen_t client_length = sizeof(client_address);
        int client_socket = accept(socket, (struct sockaddr*)&client_address, &client_length);
        if (client_socket!= -1) {
            // Handle incoming connection
            node_add_peer(node, client_socket, client_socket);
            char message[1024];
            recv(client_socket, message, 1024, 0);
            node_receive_message(node, message);
        }
    }
}

int main() {
    // Create a new node
    Node* node = node_new(1, RSA_generate_keypair(2048), RSA_generate_keypair(2048), 1);

    // Start the node
    node_start(node);

    return 0;
}

File: blockchain_integration/pi_network/qrzk-mpc/node/node.rs
// node.rs

use std::collections::{HashMap, VecDeque};
use std::net::{TcpListener, TcpStream};
use std::sync::{Arc, Mutex};
use std::thread;

use crypto::{aes, hash, rsa};
use serde::{Deserialize, Serialize};

// Define the Node struct
pub struct Node {
    id: u64,
    public_key: rsa::PublicKey,
    private_key: rsa::PrivateKey,
    network_id: u64,
    peers: HashMap<u64, TcpStream>,
    blockchain: Vec<Block>,
    mempool: VecDeque<Transaction>,
    node_state: NodeState,
}

// Define the NodeState enum
enum NodeState {
    Idle,
    Syncing,
    Mining,
}

// Define the Block struct
#[derive(Serialize, Deserialize)]
pub struct Block {
    id: u64,
    transactions: Vec<Transaction>,
    previous_hash: [u8; 32],
    timestamp: u64,
}

// Define the Transaction struct
#[derive(Serialize, Deserialize)]
pub struct Transaction {
    id: u64,
    from: [u8; 20],
    to: [u8; 20],
    amount: u64,
}

impl Node {
    // Create a new node
    pub fn new(id: u64, public_key: rsa::PublicKey, private_key: rsa::PrivateKey, network_id: u64) -> Self {
        Node {
            id,
            public_key,
            private_key,
            network_id,
            peers: HashMap::new(),
            blockchain: Vec::new(),
            mempool: VecDeque::new(),
            node_state: NodeState::Idle,
        }
    }

    // Add a peer to the node
    pub fn add_peer(&mut self, peer_id: u64, stream: TcpStream) {
        self.peers.insert(peer_id, stream);
    }

    // Remove a peer from the node
    pub fn remove_peer(&mut self, peer_id: u64) {
        self.peers.remove(&peer_id);
    }

    // Send a message to a peer
    pub fn send_message(&self, peer_id: u64, message: &str) {
        if let Some(stream) = self.peers.get(&peer_id) {
            let encrypted_message = aes::encrypt(message, &self.public_key);
            stream.write_all(&encrypted_message).unwrap();
        }
    }

    // Receive a message from a peer
    pub fn receive_message(&mut self, message: &str) {
        let decrypted_message = aes::decrypt(message, &self.private_key);
        match decrypted_message {
            "block" => {
                // Handle block message
                let block: Block = serde_json::from_str(decrypted_message).unwrap();
                self.blockchain.push(block);
            }
            "transaction" => {
                // Handle transaction message
                let transaction: Transaction = serde_json::from_str(decrypted_message).unwrap();
                self.mempool.push_back(transaction);
            }
            _ => {
                // Handle unknown message
                println!("Unknown message: {}", decrypted_message);
            }
        }
    }

    // Start the node
    pub fn start(&mut self) {
        thread::spawn(move || {
            // Start listening for incoming connections
            let listener = TcpListener::bind("0.0.0.0:8080").unwrap();
            for stream in listener.incoming() {
                match stream {
                    Ok(stream) => {
                        // Handle incoming connection
                        self.handle_incoming_connection(stream);
                    }
                    Err(e) => {
                        println!("Error: {}", e);
                    }
                }
            }
        });
    }

    // Handle incoming connection
    fn handle_incoming_connection(&mut self, stream: TcpStream) {
        // Read the peer's ID and public key
        let mut buffer = [0; 1024];
        stream.read(&mut buffer).unwrap();
        let peer_id = u64::from_le_bytes(buffer[..8].try_into().unwrap());
        let peer_public_key = rsa::PublicKey::from_bytes(&buffer[8..]);

        // Add the peer to the node
        self.add_peer(peer_id, stream);

        // Send the node's ID and public key to the peer
        let message = format!("id:{}public_key:{}", self.id, self.public_key);
        self.send_message(peer_id, &message);
    }
}

fn main() {
    // Create a new node
    let node = Node::new(1, rsa::generate_keypair().0, rsa::generate_keypair().1, 1);

    // Start the node
    node.start();
}
